[BUG] Building Apalache from source with Zulu JDK and Maven fails on MacOS

**Describe the bug**
I tried to build Apalache 0.7.2 from sources on MacOS 10.15.5, using Zulu JDK8 and Apache Maven 3.6.3. Running `make` failed while testing `tla-assignments`, because the Z3 Java library cannot be loaded. I get the following exception:
```
java.lang.UnsatisfiedLinkError: no libz3java in java.library.path
```
although both `libz3.dylib` and `libz3java.dylib` are in `/path/to/apalache/3rdparty/lib`.

**To Reproduce**
Steps to reproduce the behavior:
 1. Install Zulu JDK8. Running `java -version` outputs:
```
openjdk version "1.8.0_265"
OpenJDK Runtime Environment (Zulu 8.48.0.53-CA-macosx) (build 1.8.0_265-b11)
OpenJDK 64-Bit Server VM (Zulu 8.48.0.53-CA-macosx) (build 25.265-b11, mixed mode)
```
 2. Install Apache Maven 3.6.3. Running `mvn -v` outputs:
```
Apache Maven 3.6.3 
Maven home: /opt/apache-maven-3.6.3
Java version: 1.8.0_265, vendor: Azul Systems, Inc., runtime: /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/jre
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "10.15.5", arch: "x86_64", family: "mac"

```
 3. Running `make` in the `apalache` directory outputs:
```
...
[INFO] ----------------< at.forsyte.apalache:tla-assignments >-----------------
[INFO] Building tla-assignments 0.7.2-RELEASE                            [7/10]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ tla-assignments ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/ilina/apalache/tla-assignments/src/main/resources
[INFO] 
[INFO] --- scala-maven-plugin:4.4.0:add-source (scala-compile-first) @ tla-assignments ---
[INFO] Add Source directory: /Users/ilina/apalache/tla-assignments/src/main/scala
[INFO] Add Test Source directory: /Users/ilina/apalache/tla-assignments/src/test/scala
[INFO] 
[INFO] --- scala-maven-plugin:4.4.0:compile (scala-compile-first) @ tla-assignments ---
[INFO] Using incremental compilation using Mixed compile order
[INFO] Compiler bridge file: /Users/ilina/.sbt/1.0/zinc/org.scala-sbt/org.scala-sbt-compiler-bridge_2.12-1.3.5-bin_2.12.0__52.0-1.3.5_20200330T000931.jar
[INFO] Compiling 15 Scala sources to /Users/ilina/apalache/tla-assignments/target/classes ...
[INFO] Done compiling.
[INFO] compile in 8.5 s
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ tla-assignments ---
[INFO] Changes detected - recompiling the module!
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default) @ tla-assignments ---
[INFO] Changes detected - recompiling the module!
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ tla-assignments ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 31 resources
[INFO] 
[INFO] --- scala-maven-plugin:4.4.0:testCompile (scala-test-compile) @ tla-assignments ---
[INFO] Using incremental compilation using Mixed compile order
[INFO] Compiler bridge file: /Users/ilina/.sbt/1.0/zinc/org.scala-sbt/org.scala-sbt-compiler-bridge_2.12-1.3.5-bin_2.12.0__52.0-1.3.5_20200330T000931.jar
[INFO] Compiling 4 Scala sources to /Users/ilina/apalache/tla-assignments/target/test-classes ...
[INFO] Done compiling.
[INFO] compile in 7.6 s
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ tla-assignments ---
[INFO] Changes detected - recompiling the module!
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ tla-assignments ---
[INFO] Surefire report directory: /Users/ilina/apalache/tla-assignments/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running at.forsyte.apalache.tla.assignments.TestAssignmentStrategyEncoder
Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.549 sec
Running at.forsyte.apalache.tla.assignments.TestSymbTransPass
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.04 sec <<< FAILURE!
at.forsyte.apalache.tla.assignments.TestSymbTransPass  Time elapsed: 0.027 sec  <<< ERROR!
java.lang.UnsatisfiedLinkError: no libz3java in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1875)
	at java.lang.Runtime.loadLibrary0(Runtime.java:872)
	at java.lang.System.loadLibrary(System.java:1124)
	at com.microsoft.z3.Native.<clinit>(Native.java:14)
	at com.microsoft.z3.Context.<init>(Context.java:41)
	at at.forsyte.apalache.tla.assignments.SMTInterface.apply(SMTInterface.scala:55)
	at at.forsyte.apalache.tla.assignments.TestSymbTransPass.$anonfun$new$2(TestSymbTransPass.scala:73)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at org.scalatest.FunSuite.withFixture(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
	at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
	at org.scalatest.FunSuite.runTest(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
	at scala.collection.immutable.List.foreach(List.scala:378)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
	at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
	at org.scalatest.Suite.run(Suite.scala:1147)
	at org.scalatest.Suite.run$(Suite.scala:1129)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
	at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
	at org.scalatest.FunSuite.run(FunSuite.scala:1560)
	at org.scalatest.junit.JUnitRunner.run(JUnitRunner.scala:99)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

Running at.forsyte.apalache.tla.assignments.TestSymbTransGenerator
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.044 sec
Running at.forsyte.apalache.tla.assignments.TestAlphaTransform
Parsing file /Users/ilina/apalache/tla-assignments/src/test/resources/assignmentSolver/Paxos.tla
Parsing file /private/var/folders/r3/_hxg4k0x4sgf0vh4jkv91hvh0000gn/T/Integers.tla
Parsing file /private/var/folders/r3/_hxg4k0x4sgf0vh4jkv91hvh0000gn/T/Naturals.tla
Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.348 sec

Results :

Tests in error: 
  at.forsyte.apalache.tla.assignments.TestSymbTransPass: no libz3java in java.library.path

Tests run: 22, Failures: 0, Errors: 1, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for APALACHE project 0.7.2-RELEASE:
[INFO] 
[INFO] APALACHE project ................................... SUCCESS [  0.993 s]
[INFO] tlair .............................................. SUCCESS [01:07 min]
[INFO] infra .............................................. SUCCESS [ 24.228 s]
[INFO] tla-import ......................................... SUCCESS [ 46.659 s]
[INFO] tla-pp ............................................. SUCCESS [ 54.156 s]
[INFO] tla-types .......................................... SUCCESS [ 29.994 s]
[INFO] tla-assignments .................................... FAILURE [ 17.611 s]
[INFO] tla-bmcmt .......................................... SKIPPED
[INFO] tool ............................................... SKIPPED
[INFO] apalache-pkg ....................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:01 min
[INFO] Finished at: 2020-09-08T14:24:04+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.12.4:test (default-test) on project tla-assignments: There are test failures.
[ERROR] 
[ERROR] Please refer to /Users/ilina/apalache/tla-assignments/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :tla-assignments
make: *** [apalache] Error 1
```

**Desktop (please complete the following information):**
 - OS: Mac OS 10.15.5
 - JDK version: OpenJDK Runtime Environment (Zulu 8.48.0.53-CA-macosx)
 - Version: 0.7.2

**Additional context**
I have tried several fixes, to no avail:
 - setting the environment variable `JAVA_LIBRARY_PATH` and `DYLD_LIBRARY_PATH` to `/path/to/apalache/3rdparty/lib`
 - running `java -Djava.library.path=/path/to/apalache/3rdparty/lib`  
 - passing `-DargLine='-Djava.library.path=/path/to/apalache/3rdparty/lib'` as an argument to the `mvn` command in the Makefile
 
Building Apalache with `make build-quick` succeeds. However, the error persists when running `apalache-mc check spec.tla`.

FWIW I am also unable to compile Apalache (full log below).

I wonder whether it might be worth porting the build to sbt at some point, and see if that smoothes things out?

<details>
<summary>Click to show full log</summary>
<pre>
[INFO] ------------------------------< Box:box >-------------------------------
[INFO] Building box 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ box ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory /Users/coromac/Informal/Code/Apalache/apalache/3rdparty/Box/Box/src/main/resources
[INFO]
[INFO] --- scala-maven-plugin:3.1.6:add-source (scala-compile-first) @ box ---
[INFO] Add Source directory: /Users/coromac/Informal/Code/Apalache/apalache/3rdparty/Box/Box/src/main/scala
[INFO] Add Test Source directory: /Users/coromac/Informal/Code/Apalache/apalache/3rdparty/Box/Box/src/test/scala
[INFO]
[INFO] --- scala-maven-plugin:3.1.6:compile (scala-compile-first) @ box ---
[WARNING]  Expected all dependencies to require Scala version: 2.12.0
[WARNING]  org.scalatest:scalatest_2.12:3.0.4 requires scala version: 2.12.3
[WARNING] Multiple versions of scala libraries detected!
[INFO] /Users/coromac/Informal/Code/Apalache/apalache/3rdparty/Box/Box/src/main/scala:-1: info: compiling
[INFO] Compiling 3 source files to /Users/coromac/Informal/Code/Apalache/apalache/3rdparty/Box/Box/target/classes at 1600251995957
[ERROR] error: scala.reflect.internal.MissingRequirementError: object java.lang.Object in compiler mirror not found.
[ERROR] 	at scala.reflect.internal.MissingRequirementError$.signal(MissingRequirementError.scala:17)
[ERROR] 	at scala.reflect.internal.MissingRequirementError$.notFound(MissingRequirementError.scala:18)
[ERROR] 	at scala.reflect.internal.Mirrors$RootsBase.$anonfun$getModuleOrClass$4(Mirrors.scala:54)
[ERROR] 	at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:54)
[ERROR] 	at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:45)
[ERROR] 	at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:45)
[ERROR] 	at scala.reflect.internal.Mirrors$RootsBase.getModuleOrClass(Mirrors.scala:66)
[ERROR] 	at scala.reflect.internal.Mirrors$RootsBase.getClassByName(Mirrors.scala:102)
[ERROR] 	at scala.reflect.internal.Mirrors$RootsBase.getRequiredClass(Mirrors.scala:105)
[ERROR] 	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass$lzycompute(Definitions.scala:267)
[ERROR] 	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass(Definitions.scala:267)
[ERROR] 	at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1447)
[ERROR] 	at scala.tools.nsc.Global$Run.<init>(Global.scala:1126)
[ERROR] 	at scala.tools.nsc.MainClass.doCompile(Main.scala:24)
[ERROR] 	at scala.tools.nsc.Driver.process(Driver.scala:55)
[ERROR] 	at scala.tools.nsc.Driver.main(Driver.scala:68)
[ERROR] 	at scala.tools.nsc.Main.main(Main.scala)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
[ERROR] 	at scala_maven_executions.MainHelper.runMain(MainHelper.java:164)
[ERROR] 	at scala_maven_executions.MainWithArgsInFile.main(MainWithArgsInFile.java:26)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.620 s
[INFO] Finished at: 2020-09-16T12:26:36+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.1.6:compile (scala-compile-first) on project box: wrap: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
</pre>
</summary>

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,20 +4,20 @@ matrix:
   include:
     - os: linux
       jdk: openjdk8
-      
+
 # first, fix the issues with the dynamic libraries
 #    - os: osx
 #      osx_image: xcode9.3 # the latest image compatible with JDK8
 
 python:
-    - "2.7"
+  - "2.7"
 
 # required to compile z3
 addons:
   apt:
     packages:
-    - gcc-4.7
-    - g++-4.7
+      - gcc-4.7
+      - g++-4.7
 
 cache:
   directories:
@@ -26,10 +26,8 @@ cache:
 
 # locally compile z3 and make it available via LD_LIBRARY_PATH
 before_install:
-    - pushd 3rdparty
-    - ./install-local.sh
-    - popd
-    #- pip install release-me
+  - make deps
+  #- pip install release-me
 
 #after_success:
 #- sudo pip install --upgrade pip
@@ -42,12 +40,10 @@ before_install:
 #  - unstable
 
 script:
-    - make integration
+  - make integration
 
 notifications:
-    email:
-      recipients:
-        apalache-build@forsyte.at andrey@informal.systems
-      on_success: change
-      on_failure: always
-
+  email:
+    recipients: apalache-build@forsyte.at andrey@informal.systems
+    on_success: change
+    on_failure: always

--- a/3rdparty/README.md
+++ b/3rdparty/README.md
@@ -1,15 +1,9 @@
-Third-party libraries
-=====================
+# Third-party libraries
 
 APALACHE requires the following libraries that are not available from
 the central Maven repository:
 
- * TLA+2 Tools: tla2tools.jar. Available from:
-    [TLA+ Tools page](https://lamport.azurewebsites.net/tla/tools.html "TLA+ Tools").
-
- * Z3 Java bindings and Z3 shared libraries. The source code is available
- from [Z3 Github page](https://github.com/Z3Prover/z3 "Z3 Github page").
-   Check below how one to compile Z3 libraries.
+- Box: Available from [GitHub](https://github.com/Kukovec).
 
 In order to download, build and use the above libraries,
 execute the following shell command:
@@ -18,16 +12,14 @@ execute the following shell command:
 $ ./install-local.sh
 ```
 
-The above command downloads and compiles z3, downloads TLA2Tools and
-installs the both libraries in your local Maven
-cache. Once it is done, you do not have to worry about these libraries
-anymore, they will be automatically retrieved by our build system.
+The command downloads and compiles the libraries and installs them in your
+local Maven cache. Once it is done, you do not have to worry about these
+libraries anymore, they will be automatically retrieved by our build system.
 
-After the command has finished, add the following line to
-your ```~/.bashrc``` or ```~/.zshrc```:
+Unless you are using `direnv`, as recommended in
+[CONTRIBUTING.md](../CONTRIBUTING.md), then add the following line to your
+`~/.bashrc` or `~/.zshrc`:
 
 ```
 export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$APALACHE/3rdparty
 ```
-
-(The above command is needed to load the Z3 shared libraries.)

--- a/3rdparty/install-local.sh
+++ b/3rdparty/install-local.sh
@@ -7,7 +7,6 @@
 set -e
 
 D=`dirname $0` && D=`cd $D; pwd`
-Z3_DIR="$D/z3-4.8.7"
 CACHE=$1 # either empty, or --nocache
 
 if [[ "$OSTYPE" == "darwin"* ]]; then
@@ -16,26 +15,6 @@ if [[ "$OSTYPE" == "darwin"* ]]; then
 else
     echo "Assuming that you are using Linux..."
     OST="linux"
-fi
-
-if [ "$CACHE" == "--nocache" ]; then
-    echo "Cleaning the build cache"
-    rm -rf "${Z3_DIR}" "$D/Box"
-fi
-
-if [ -f "${Z3_DIR}/configure" ]; then
-    echo "Using a cached Z3 build..."
-else
-    echo "Checking out z3..."
-    git clone https://github.com/Z3Prover/z3.git ${Z3_DIR}
-    pushd ${Z3_DIR}
-    git checkout z3-4.8.7
-    echo "Configuring z3 locally (Linux)..."
-    python scripts/mk_make.py --java -p $D/
-    echo "Compiling z3..."
-    cd build
-    make
-    popd
 fi
 
 BOXHASH=77feaf4fc873990b72b31273985e2c88b8f2f502
@@ -48,14 +27,6 @@ pushd $D/Box
 git reset --hard $BOXHASH
 popd
 
-# install Z3 libraries
-echo "Compiling and installing z3..."
-pushd ${Z3_DIR}
-cd build
-make install # install *.so and *.jar in 3rdparty
-popd
-echo "Done with z3"
-
 echo "Installing Box..."
 pushd $D/Box/Box
 mvn install
@@ -63,21 +34,11 @@ cp target/box-1.0-SNAPSHOT.jar $D/lib/box.jar
 popd
 echo "Done with Box"
 
-echo "Installing Z3 in your local maven cache..."
-
-mvn -f $D/z3-pom.xml install install:install-file \
-    "-Dfile=$D/lib/com.microsoft.z3.jar" "-DpomFile=$D/z3-pom.xml"
-
 #mvn -f $D/tla2tools-pom.xml install install:install-file \
 #    "-Dfile=$D/tla2tools.jar" "-DpomFile=$D/tla2tools-pom.xml"
 
 mvn -f $D/box-pom.xml install install:install-file \
     "-Dfile=$D/lib/box.jar" "-DpomFile=$D/box-pom.xml"
-
-if [ "$CACHE" == "--nocache" ]; then
-    echo "Cleaning the build cache"
-    rm -rf "${Z3_DIR}" "$D/Box"
-fi
 
 cat <<EOF
 1. To build Apalache, just use make.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,9 @@
 ## Unreleased #unstable
 
  * Use a staged docker build, reducing container size ~70%, see #195
+ * Use [Z3-TurnKey](https://github.com/tudo-aqua/z3-turnkey) instead of a
+   bespoke Z3 build, see #219
+ * Use Z3 version 4.8.7.1, see #219
 
 ## 0.7.2 [RELEASE]
 

--- a/Makefile
+++ b/Makefile
@@ -15,11 +15,13 @@ QUICK_MAVEN_OPTS := "-XX:+TieredCompilation -XX:TieredStopAtLevel=1 -Xverify:non
 # - run up to 4 threads per core (4C): https://cwiki.apache.org/confluence/display/MAVEN/Parallel+builds+in+Maven+3
 QUICK_MAVEN_ARGS := -DskipTests -Dscoverage.skip=true -T 4C
 
-.PHONY: all apalache compile build-quick test integration clean
+.PHONY: all apalache compile build-quick test integration clean deps
 
 all: apalache
 
-apalache: $(DEPS)
+deps: $(DEPS)
+
+apalache: deps
 	# tell maven to load the binary libraries and build the package
 	$(ENV) mvn package
 

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 # a good old Makefile for the end users, as Maven is too much pain
 
 DEPDIR=3rdparty
-DEPS=$(DEPDIR)/lib/com.microsoft.z3.jar $(DEPDIR)/lib/box.jar
+DEPS=$(DEPDIR)/lib
 ENV=JAVA_LIBRARY_PATH="$(abspath $(DEPDIR)/lib)" NO_MVN=1 LD_LIBRARY_PATH="$(abspath $(DEPDIR)/lib)"
 
 # See https://www.jrebel.com/blog/how-to-speed-up-your-maven-build
@@ -43,11 +43,7 @@ integration: apalache
 clean:
 	mvn clean
 
-$(DEPDIR)/lib/com.microsoft.z3.jar:
-	# install microsoft z3
-	cd "$(DEPDIR)" && ./install-local.sh
-
-$(DEPDIR)/lib/box.jar:
+$(DEPDIR)/lib:
+	mkdir -p $(DEPDIR)/lib
 	# install box by Jure (fix in the future!)
 	cd "$(DEPDIR)" && ./install-local.sh
-

--- a/bin/apalache-mc
+++ b/bin/apalache-mc
@@ -40,11 +40,12 @@ TLA_LIB="$DIR/src/tla"
 # Once tlaplus/#490,#493 are released, remove this workaround.
 (test -z "$TLA_PATH" && TLA_PATH="$TLA_LIB") || TLA_PATH="$TLA_LIB:$TLA_PATH"
 
-# set LD_LIBRARY_PATH to find z3 libraries
 export LD_LIBRARY_PATH="$DIR/3rdparty/lib:${LD_LIBRARY_PATH}"
+
 # 1. The maximum heap size, Z3 will use much more as a native library.
 # 2. Duplicating LD_LIBRARY_PATH, tell JVM to find the third-party libraries.
 JVM_ARGS="-Xmx4096m -Djava.library.path=$DIR/3rdparty/lib "
+
 # uncomment to track memory usage with: jcmd <pid> VM.native_memory summary
 #JVM_ARGS="${JVM_ARGS} -XX:NativeMemoryTracking=summary"
 

--- a/docs/manual.md
+++ b/docs/manual.md
@@ -194,8 +194,8 @@ $ docker image build -t apalache:0.7.0 .
    - On Arch: `sudo pacman -Syu maven`
 4. Clone the git repository: `git clone https://github.com/informalsystems/apalache.git`.
 5. Change into the project directory: `cd apalache`.
-7. Run `make`. This command will install Microsoft Z3, compile Apalache
-   and assemble the package.
+7. Run `make`. This command will install third party dependencies, compile
+   Apalache, and assemble the package.
 6. *Optionally* install [direnv][] and run `direnv allow`
 8. Confirm you can run the executable. It should print the inline CLI help message.
    - If you used `direnv`, then `apalache-mc` will be in your path.

--- a/pom.xml
+++ b/pom.xml
@@ -126,12 +126,12 @@
                 <version>1.6.0-SNAPSHOT</version>
             </dependency>
 
-            <!-- Dependencies that are not available from the Maven repository, see 3rdparty/README.md -->
-
+            <!-- Z3-TurnKey provides Z3 and the integration of its Java API in
+                 a portable way. See https://github.com/tudo-aqua/z3-turnkey -->
             <dependency>
-                <groupId>com.microsoft</groupId>
-                <artifactId>z3</artifactId>
-                <version>4.8.7</version>
+                <groupId>io.github.tudo-aqua</groupId>
+                <artifactId>z3-turnkey</artifactId>
+                <version>4.8.7.1</version>
             </dependency>
         </dependencies>
     </dependencyManagement>

--- a/tla-assignments/pom.xml
+++ b/tla-assignments/pom.xml
@@ -46,10 +46,6 @@
             <artifactId>tla2tools</artifactId>
         </dependency>
         <dependency>
-            <groupId>com.microsoft</groupId>
-            <artifactId>z3</artifactId>
-        </dependency>
-        <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
             <scope>test</scope>
@@ -58,6 +54,10 @@
             <groupId>org.scalatest</groupId>
             <artifactId>scalatest_${scalaBinaryVersion}</artifactId>
             <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.github.tudo-aqua</groupId>
+            <artifactId>z3-turnkey</artifactId>
         </dependency>
     </dependencies>
 </project>

--- a/tla-bmcmt/pom.xml
+++ b/tla-bmcmt/pom.xml
@@ -52,8 +52,8 @@
             <artifactId>tla2tools</artifactId>
         </dependency>
         <dependency>
-            <groupId>com.microsoft</groupId>
-            <artifactId>z3</artifactId>
+            <groupId>io.github.tudo-aqua</groupId>
+            <artifactId>z3-turnkey</artifactId>
         </dependency>
         <dependency>
             <groupId>junit</groupId>

--- a/tla-import/pom.xml
+++ b/tla-import/pom.xml
@@ -47,8 +47,8 @@
             <artifactId>tla2tools</artifactId>
         </dependency>
         <dependency>
-            <groupId>com.microsoft</groupId>
-            <artifactId>z3</artifactId>
+            <groupId>io.github.tudo-aqua</groupId>
+            <artifactId>z3-turnkey</artifactId>
         </dependency>
         <dependency>
             <groupId>junit</groupId>

--- a/tla-pp/pom.xml
+++ b/tla-pp/pom.xml
@@ -42,8 +42,8 @@
             <artifactId>tla2tools</artifactId>
         </dependency>
         <dependency>
-            <groupId>com.microsoft</groupId>
-            <artifactId>z3</artifactId>
+            <groupId>io.github.tudo-aqua</groupId>
+            <artifactId>z3-turnkey</artifactId>
         </dependency>
         <dependency>
             <groupId>junit</groupId>

--- a/tla-types/pom.xml
+++ b/tla-types/pom.xml
@@ -40,8 +40,8 @@
             <artifactId>tla2tools</artifactId>
         </dependency>
         <dependency>
-            <groupId>com.microsoft</groupId>
-            <artifactId>z3</artifactId>
+            <groupId>io.github.tudo-aqua</groupId>
+            <artifactId>z3-turnkey</artifactId>
         </dependency>
         <dependency>
             <groupId>junit</groupId>

